1_fetch.smk

```diff
@@ -0,0 +1,43 @@
+import os
+
+configfile: "1_fetch/fetch_config.yaml"
+
+
+def get_mntoha_input_files(wildcards):
+    '''List all MNTOHA files to download
+    Includes metadata, temperature observations, and drivers
+
+    :param wildcards: Snakemake wildcards (unused)
+    :returns: list of filepaths to be downloaded
+
+    '''
+    out_dir = "1_fetch/out"
+    categories = config["mntoha"].keys()
+    input_files = []
+    # include files under files key in each category
+    for category in categories:
+        for filename in config["mntoha"][category]["files"]:
+            input_files.append(os.path.join(out_dir, category + "_mntoha", filename))
+    # include driver files: 3 types x 13 suffixes = 39 files
+    for driver_type in config["mntoha"]["drivers"]["driver_types"]:
+        for suffix in config["mntoha"]["drivers"]["suffixes"]:
+            input_files.append(os.path.join(out_dir, "drivers_mntoha", f"{driver_type}_{suffix}.zip"))
+    return input_files
+
+
+rule fetch_all:
+    input:
+        "1_fetch/in/pull_date.txt",
+        get_mntoha_input_files
+
+
+rule fetch_mntoha_data_file:
+    input:
+        "1_fetch/in/pull_date.txt"
+    output:
+        "1_fetch/out/{file_category}_mntoha/{file}"
+    params:
+        sb_id = lambda wildcards: config["mntoha"][wildcards.file_category]["sb_id"]
```

Comment:
Just trying to wrap my head around what `snakemake` is doing here. So this works b/c `snakemake` automatically uses the `mntoha` dictionary from `1_fetch/fetch_config.yml` as `wildcards` and then the keys from that dictionary as `file_category`?

Reply:
Okay, here's my understanding. So, if the rule `fetch_mntoha_data_file` is being called, that means that `snakemake` has matched a requested file path to the pattern defined here in `output`. [`wildcards`](https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#wildcards) contains both of the wildcards that were named in the pattern defined in `output`: `file_category` and `file`. If a function is called from within a `snakemake` rule, the first argument passed to that function is `wildcards`. That goes for both named functions (e.g. `get_mntoha_input_files`) and `lambda` functions like the one in line 40. That's why `wildcards` is named the same in lines 6 and 40 - it's that argument - even though in line 6, we know it's the `wildcards` from the `fetch_all` rule, while in line 40, it's the `wildcards` from the `fetch_mntoha_data_file` rule. From there, like you said, `snakemake` gets the `sb_id` by indexing the `config` that's loaded from `1_fetch/fetch_config.yaml`.

Reply:
Alternatively, we could have defined a function, like:
```
def get_sb_id(wildcards):
    file_category = wildcards.file_category
    return config["mntoha"][file_category]["sb_id"]

rule fetch_mntoha_data_file:
    input:
        "1_fetch/in/pull_date.txt"
    output:
        "1_fetch/out/{file_category}_mntoha/{file}"
    params:
        sb_id = get_sb_id
    script:
        "1_fetch/src/sb_fetch.py"


```

Reply:
Thanks, Andy - that makes complete sense now.

Reply:
I agree with your explanation, @AndyMcAliley.

I'll add a concrete example, in case that is helpful. 

- Snakemake always works from what you are asking it to make. In `fetch_all`, one of the things we are asking it to make is `1_fetch/out/drivers/input_cells.png` (one file in the list of many files that the `get_mntoha_input_files` function returns).
- Snakemake finds the rule that makes `1_fetch/out/drivers/input_cells.png`. This is `fetch_mntoha_data_file` rule. 
- The `wildcards` object gets built by parsing the filename according to the wildcards defined in the `output` string of the rule (`"1_fetch/out/{file_category}_mntoha/{file}"`). 
- So in this case `wildcards.file_type = "drivers"` and `wildcards.file = "input_cells.png"`. 
- Then, the `params` are defined. In this case, the one-line `lambda` function will return `config["mntoha"]["drivers"]["sb_id"]` which is `"5e5d0b96e4b01d50924f2b34"` from the config file. 
- That `sb_id` is then accessed via the `params` object in the `1_fetch/src/sb_fetch.py` file since you are using the `script` functionality.

Reply:
That is helpful -- thanks, Jeff!